### PR TITLE
#94 heroic game roots

### PR DESCRIPTION
**Pull request by sluedecke**

### CODE REVIEW

File: src/heroic.rs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    config::RootsConfig,
+    manifest::{Manifest, Store},
+    prelude::{normalize_title, StrictPath},
+};
+
+// Deserialization of Heroic gog_store/installed.json
+#[derive(serde::Deserialize)]
+struct HeroicInstalledGame {
+    #[serde(rename = "appName")]
+    game_id: String,
```
Comment (mtkennerly) [inline src/heroic.rs:13]: Adapting from camel case makes sense, but I think we should keep the name itself for consistency:

```rust
    #[serde(rename = "appName")]
    app_name: String,
```

File: src/heroic.rs [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+#[derive(serde::Deserialize)]
+struct GogLibrary {
+    games: Vec<GogLibraryGame>,
+}
+
+// Deserialization of Legendary legendary/installed.json
+#[derive(serde::Deserialize)]
+struct LegendaryInstalledGame {
+    #[serde(rename = "app_name")]
+    game_id: String,
```
Comment (mtkennerly) [inline src/heroic.rs:36]: Let's just stick to:

```rust
    app_name: String,
```

File: src/heroic.rs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    title: String,
+    platform: String,
+}
+#[derive(serde::Deserialize)]
+struct LegendaryInstalled(HashMap<String, LegendaryInstalledGame>);
+
+// TODO.2022-10-10 heroic: windows location for legendary
```
Comment (mtkennerly) [inline src/heroic.rs:43]: It's the same :)

File: src/heroic.rs [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+// TODO.2022-10-15 heroic: make relative to heroic root in question!
```
Comment (mtkennerly) [inline src/heroic.rs:44]: Does this comment just apply to the Flatpak path?

File: src/heroic.rs [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+        let mut instance = HeroicGames {
+            normalized_to_official: manifest
+                .0
+                .keys()
+                .map(|title| (normalize_title(title), title.clone()))
+                .collect(),
+            ..Default::default()
+        };
+
+        roots.iter().for_each(|root: &RootsConfig| {
```
Comment (mtkennerly) [inline src/heroic.rs:72]: I prefer regular for-loops over `for_each`. If you just needed to avoid an ownership issue, you can borrow:

```
for root in &roots {}
```

File: src/heroic.rs [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+        for &legendary_path_candidate in LEGENDARY_PATHS {
+            let legendary_path = StrictPath::new(legendary_path_candidate.to_string());
+            if legendary_path.is_dir() {
+                log::trace!(
+                    "detect_legendary_roots found legendary configuration in {}",
+                    legendary_path.interpret()
+                );
+
+                let mut legendary_installed = legendary_path.as_std_path_buf();
+                legendary_installed.push("installed.json");
```
Comment (mtkennerly) [inline src/heroic.rs:101]:
```rust
let legendary_installed = legendary_path.joined("installed.json");
```

File: src/heroic.rs [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+        // TODO.2022-10-15 heroic: handle games storing stuff in ~/.var/app/com.heroicgameslauncher.hglw
+        // TODO.2022-10-15 heroic: for some games, .var/app/com.heroicgameslauncher.hglw is a root just like $HOME
+    }
+
+    fn detect_legendary_roots(&mut self, root: &RootsConfig) {
+        log::trace!("detect_legendary_roots searching for legendary config...");
+
+        for &legendary_path_candidate in LEGENDARY_PATHS {
+            let legendary_path = StrictPath::new(legendary_path_candidate.to_string());
+            if legendary_path.is_dir() {
```
Comment (mtkennerly) [inline src/heroic.rs:94]: One trick for flattening the code is to do an early return/continue on the error case. That way, you can turn this:

```
if legendary_path.is_dir() {
    // indent grows
} else {
    // log
}
```

...into this:

```
if !legendary_path.is_dir() {
    // log
    continue;
}

// indent doesn't grow
```

When you have several error checks chained together like this, it makes the code easier to read.

File: src/heroic.rs [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+        );
+
+        // use gog_store/library.json to build map .app_name -> .title
+        let mut game_titles = std::collections::HashMap::<String, String>::new();
+        let gog_library = serde_json::from_str::<GogLibrary>(
+            &std::fs::read_to_string(format!("{}/gog_store/library.json", root.path.interpret())).unwrap_or_default(),
+        );
+        gog_library.unwrap().games.iter().for_each(|game| {
+            game_titles.insert(game.app_name.clone(), game.title.clone());
+        });
```
Comment (mtkennerly) [inline src/heroic.rs:143]: Instead of unwrapping, we can return early if it fails.

Also, we should be able to do something like:

```rust
let game_titles: HashMap<String, String> = gog_library.games.iter().map(|game| {
    (game.app_name.clone(), game.title.clone())
}).collect();
```

File: src/heroic.rs [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+            "detect_gog_roots found {} games in {}",
+            game_titles.len(),
+            format!("{}/gog_store/library.json", root.path.interpret())
+        );
+
+        // iterate over all games found in HEROCONFIGDIR/gog_store/installed.json and call find_game_root
+        let content = std::fs::read_to_string(format!("{}/gog_store/installed.json", root.path.interpret()));
+        if let Ok(installed_games) = serde_json::from_str::<HeroicInstalled>(&content.unwrap_or_default()) {
+            installed_games.installed.iter().for_each(|game| {
+                let game_title = game_titles.get(&game.game_id).unwrap();
```
Comment (mtkennerly) [inline src/heroic.rs:154]: I think we should continue to the next iteration instead of unwrapping. This probably shouldn't fail, but it's easy enough to avoid the unwrap.

File: src/heroic.rs [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+                if let Some(sp) = self.find_game_root(root.path.interpret(), game_title, &game.platform, &game.game_id)
+                {
+                    self.memorize_game_root(game_title, &sp);
+                }
+            })
+        }
+    }
+
+    fn memorize_game_root(&mut self, title: &String, path: &StrictPath) {
```
Comment (mtkennerly) [inline src/heroic.rs:163]: I think there are some places here where the term `root` is used when it should be `prefix`.

**Review by mtkennerly — CHANGES_REQUESTED** [review]
Thanks for working on this! There are some things I'd like changed, but this is a great start.

**Comment by mtkennerly** [comment]
Linking #94 for discoverability. Doesn't seem to want to link from the title.

**Comment by sluedecke** [comment]
Great, thank you for thoroughly looking into it!  I will work through the list of changes.

File: src/config.rs [hunk 11]
```diff
@@ -26,6 +26,8 @@ pub struct Config {
     pub restore: RestoreConfig,
     #[serde(default, rename = "customGames")]
     pub custom_games: Vec<CustomGame>,
+    #[serde(skip)]
+    pub heroic_roots: std::collections::HashMap<String, StrictPath>,
```
[the remark this answers is not in the document]
Comment (sluedecke) [reply src/config.rs:30]: Nice find, I missed that completely - it is a leftover.  Since it was set to an empty map in the test code, the compiler did not complain.

File: src/heroic.rs (hunk 1 above)
Comment (sluedecke) [reply src/heroic.rs:13]: The reason I changed it is that `app_name` actually contains a game id.  This is consistent across all heroic and legendary json files so either name seems to be ok to me.  Shall I keep it as in the json or do you prefer a more content oriented variable name?

File: src/heroic.rs (hunk 2 above)
Comment (sluedecke) [reply src/heroic.rs:36]: same as comment above

File: src/heroic.rs (hunk 4 above)
Comment (sluedecke) [reply src/heroic.rs:44]: The relative path? On my systems it applies to both kinds of installations.  Normal is `~/.config/heroic` + `~/.config/legendary` and flatpak is `~/.var/app/com.heroicgameslauncher.hgl/config/heroic` + `~/.var/app/com.heroicgameslauncher.hgl/config/legendary`

Since heroic refers to it's sibbling legendary installation a relative path seems to be a good idea to make sure ludusavi finds the correct one.

File: src/heroic.rs (hunk 5 above)
Comment (sluedecke) [reply src/heroic.rs:72]: Yes, much better and concise style!  I've read the rust book chapter on loops after I've written this line :)

File: src/heroic.rs (hunk 7 above)
Comment (sluedecke) [reply src/heroic.rs:94]: It [does indeed](https://enterprisecraftsmanship.com/posts/early-exit-is-tail-call-optimization/)

File: src/heroic.rs (hunk 8 above)
Comment (sluedecke) [reply src/heroic.rs:143]: Replaced all unwrap() with checks for Some

File: src/heroic.rs (hunk 9 above)
Comment (sluedecke) [reply src/heroic.rs:154]: Checking for Some()

File: src/config.rs [hunk 12]
```diff
@@ -531,6 +533,14 @@ impl Config {
             // GOG Galaxy:
             (format!("{}/GOG Galaxy/Games", pf32), Store::GogGalaxy),
             (format!("{}/GOG Galaxy/Games", pf64), Store::GogGalaxy),
+            // Heroic Config:
+            // TODO.2022-10-07 heroic: add default windows locations for heroic, probably similar to GOG Galaxy
```
[the remark this answers is not in the document]
Comment (sluedecke) [reply src/config.rs:537]: Let's add proper windows support in a separate item then.  This way we gain (basic) Linux Heroic support here and extend it in #142 and a new item.

File: src/heroic.rs (hunk 1 above)
Comment (mtkennerly) [reply src/heroic.rs:13]: I agree that "ID" fits better than "name", but since we're dealing with an external interface, I'd like to use it as-is. At least for me, it just makes it a little easier to cross-reference with the file. We could always add a doc comment:

```rust
    /// This is an opaque ID, not the human-readable title.
    #[serde(rename = "appName")]
    app_name: String,
```

File: src/heroic.rs (hunk 4 above)
Comment (mtkennerly) [reply src/heroic.rs:44]: Gotcha. If we do a relative/sibling check, we would still need the absolute path as a fallback. At least on Windows, they're not sibling folders. We can handle this later in #142 and #144.

File: src/heroic.rs [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+                        Some(StrictPath::new(format!(
+                            "{}/pfx",
+                            v[&game_id]["winePrefix"].as_str().unwrap_or_default()
+                        )))
+                    }
+
+                    _ => {
+                        // TODO.2022-10-07 handle unknown wine types, lutris?
+                        log::warn!(
+                            "find_game_root found Heroic Windows game {} ({}), checking... unknown wine_type: {:#?}",
```
[the remark this answers is not in the document]
Comment (mtkennerly) [reply src/heroic.rs]: Oh, sorry, I misinterpreted the log message and was thinking it was something that would always log if you ran Ludusavi on Windows.

Still, even if we don't support Lutris prefixes for now, we can just document that instead of logging a warning. After the Heroic implementation is done, I can update the README's roots section with a note about what we do and don't support.

File: src/heroic.rs [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+        &self,
+        heroic_path: String,
+        game_name: &String,
+        platform: &String,
+        game_id: &String,
+    ) -> Option<StrictPath> {
+        match platform.as_str() {
+            "windows" => {
+                // no struct for type safety used here since GamesConfig use the game id as a key name
+                let v: serde_json::Value = serde_json::from_str(
```
[the remark this answers is not in the document]
Comment (sluedecke) [reply src/heroic.rs:185]: I researched this and it would be the correct approach, if only the json would be "well" formed.  `GameConfig/*.json` look like:

```json
{
  "1207663333": {
    "autoInstallDxvk": false,
    "autoInstallVkd3d": false,
    "preferSystemLibs": false,
    "nvidiaPrime": false,
    "enviromentOptions": [],
    "wrapperOptions": [
      {
        "exe": "/home/MYUSER/Projekte/workspace/ludusavi-cloud/ludusavi-cloud.sh"
        "args": ""
      }
    ],
    "showFps": false,
    "useGameMode": false,
    "language": "",
    "wineVersion": {
      "bin": "/usr/bin/wine",
      "name": "Wine Default - wine-ge-7.0 (Staging)",
      "type": "wine",
      "wineserver": "/usr/bin/wineserver",
      "wineboot": "/usr/bin/wineboot"
    },
    "winePrefix": "/home/MYUSER/Games/Heroic/Prefixes/Avernum"
  },
  "version": "v0",
  "explicit": true
}
```

So the first maps a dynamic key to a GamesConfig, the others map to string.  I could make the wrapper a `struct GamesConfigWrapper(HashMap<String, String>)` and then parse the value associated with of the `app_name` to a `GamesConfig`.  That would be a compromise.

[the remark this answers is not in the document]
Comment (mtkennerly) [reply src/heroic.rs:185]: Ah, yeah, that's inconvenient. If you put the value type as `String`, I'd expect it to reject/discard the object and boolean values.

For cases like this, you can use an enum to preserve type-safety:

```rust
#[derive(serde::Deserialize)]
struct GamesConfigWrapper(HashMap<String, GamesConfig>);

#[derive(serde::Deserialize)]
enum GamesConfig {
    Standard {
        // winePrefix, etc
    }
    #[serde(other)]
    Other
}
```

If we ever needed to check the `version` field, for example, we could add another variant like `Text(String)`.

[the remark this answers is not in the document]
Comment (sluedecke) [reply src/heroic.rs:185]: Indeed it rejected it, kindly telling me that `find_prefix error invalid type: map, expected a string at line 2 column 16, ignoring`.

The solution you recommend does not (yet?) work either.  Serde untagged does not like Other and fails compilation with `error: #[serde(other)] cannot appear on untagged enum`.

So I tried this:

```rust
#[derive(serde::Deserialize, Debug)]
struct GamesConfigWrapper(HashMap<String, GamesConfig>);

#[derive(serde::Deserialize, Debug)]
#[serde(untagged)]
enum GamesConfig {
    Config {
        #[serde(rename = "winePrefix")]
        wine_prefix: String,
        #[serde(rename = "wineVersion")]
        wine_version: GamesConfigWine,
    },
    Version {
        version: String,
    },
    Explicit {
        explicit: bool,
    },
}
```
It compiles, but does not like the json yet:  `find_prefix error: 'data did not match any variant of untagged enum GamesConfig at line 26 column 17', ignoring`.

[the remark this answers is not in the document]
Comment (mtkennerly) [reply src/heroic.rs:185]: Could you try taking my version and just removing the `#[serde(other)]` line?

[the remark this answers is not in the document]
Comment (sluedecke) [reply src/heroic.rs:185]: Got it.  A working solution is this:

```rust
#[derive(serde::Deserialize, Debug)]
struct GamesConfigWrapper(HashMap<String, GamesConfig>);

#[derive(serde::Deserialize, Debug)]
#[serde(untagged)]
enum GamesConfig {
    Config {
        #[serde(rename = "winePrefix")]
        wine_prefix: String,
        #[serde(rename = "wineVersion")]
        wine_version: GamesConfigWine,
    },
    Version(String),
    Explicit(bool),
}

#[derive(serde::Deserialize, Debug)]
struct GamesConfigWine {
    #[serde(rename = "type")]
    wine_type: String,
}
```

[the remark this answers is not in the document]
Comment (sluedecke) [reply src/heroic.rs:185]: Better solution ignoring everything except the wanted data structure:

```rust
#[derive(serde::Deserialize, Debug)]
struct GamesConfigWrapper(HashMap<String, GamesConfig>);

#[derive(serde::Deserialize, Debug)]
#[serde(untagged)]
enum GamesConfig {
    Config {
        #[serde(rename = "winePrefix")]
        wine_prefix: String,
        #[serde(rename = "wineVersion")]
        wine_version: GamesConfigWine,
    },
    Other(serde::de::IgnoredAny),
}

#[derive(serde::Deserialize, Debug)]
struct GamesConfigWine {
    #[serde(rename = "type")]
    wine_type: String,
}
```

File: src/heroic.rs (hunk 1 above)
Comment (sluedecke) [reply src/heroic.rs:13]: Doc comment I take then.

File: src/heroic.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (sluedecke) [reply src/heroic.rs]: I installed "Alba: A Wildlife Adventure" and chose a lutris variant of wine.  It uses the same structure as a plain vanilla wine, so this is not a special case to be handled.

I recommend keeping the log::warn! so that we might catch it in the logs instead of crashing ludusavi if ever GameConfig introduces a new type.  What do you think?

File: src/heroic.rs [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+// TODO.2022-10-15 heroic: make relative to heroic root in question!
+const LEGENDARY_PATHS: &[&str] = &[
+    "~/.config/legendary",
+    // TODO.2022-10-20 heroic: flatpak install is not supported yet
+    // "~/.var/app/com.heroicgameslauncher.hgl/config/legendary",
+];
+
+#[derive(Clone, Default, Debug)]
+pub struct HeroicGames {
+    games: HashMap<String, StrictPath>,
```
[the remark this answers is not in the document]
Comment (sluedecke) [reply src/heroic.rs:53]: I agree.  Speaking from my experience with having two installations of heroic by accident (didn't properly remove one of them) this causes confusion.

[the remark this answers is not in the document]
Comment (sluedecke) [reply src/heroic.rs:53]: I wonder: if someone does that (switch from standalone to flatpak or vice versa), how should ludusavi react on restore?

[the remark this answers is not in the document]
Comment (mtkennerly) [reply src/heroic.rs]: The user could set up a redirect to migrate during restoration if they'd like. Ludusavi would still restore to the old location by default, though, since we don't want to move files around unless the user asks for it.

File: src/heroic.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (mtkennerly) [reply src/heroic.rs]: Personally, I feel like we should only use warning when something has probably gone wrong. If they add a new type and we just don't support it yet, then that's a missing feature, but not a failure. If we can handle the three currently known cases, then I'd be fine with `info` for a new unknown case, though.

**Review by mtkennerly — CHANGES_REQUESTED** [review]
Other than the tests not compiling, I think this is very close to done.

File: src/heroic.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (sluedecke) [reply src/heroic.rs]: Though an edge case, it might be worth documenting it somewhere.  Do you plan to add a wiki to the repository or would the README.md a better place for that?

File: src/heroic.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (sluedecke) [reply src/heroic.rs]: reduced to a `log::info!`

File: src/heroic.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (mtkennerly) [reply src/heroic.rs]: I can add a note to the README. There's already a section on redirects, so I could add a portion to cover this specific kind of use case.

[the remark this answers is not in the document]
Comment (sluedecke) [reply src/heroic.rs]: sounds good!

File: src/heroic.rs [hunk 16]
```diff
@@ -69,179 +107,185 @@ impl HeroicGames {
...
-                    }
+                ) {
+                    Ok(games_config_wrapper) => {
+                        if let Some(game_config) = games_config_wrapper.0.get(app_name) {
+                            match game_config {
+                                GamesConfig::Config {
+                                    wine_version,
+                                    wine_prefix,
+                                } => match wine_version.wine_type.as_str() {
+                                    "wine" => {
```
[the remark this answers is not in the document]
Comment (sluedecke) [reply src/heroic.rs:238]: Lutris is no special case and not marked as such in the heroic json files.  My test with ''Alba" shows that the prefix it generates does not differ from a standard wine installation.  So I do not consider this a special case.

**Comment by sluedecke** [comment]
So far all seems to be done and is merged.  I kindly request another review round - the final one?

**Review by mtkennerly — APPROVED** [review]
Good work 🎉 Thanks again for adding this.
